src/toxcore/dht/server/ping.rs

```diff
@@ -0,0 +1,236 @@
+        node.last_resp_time.elapsed() > Duration::from_secs(server.config.bad_node_timeout)
+    }
+
+    fn is_friend(node: &PackedNode, server: &Server) -> bool {
+        server.friends.read().iter().any(|friend| friend.pk == node.pk)
+    }
+
+    fn is_in_close_list(node: &PackedNode, server: &Server) -> bool {
+        server.friends.read().iter()
+            .map(|friend| friend.close_nodes.nodes.iter().any(|peer| peer.pk == node.pk))
+            .any(|result_bool| result_bool)
+    }
+
+    fn is_in_ping_list(node: &PackedNode, ping_list: &Bucket) -> bool {
+        ping_list.nodes.iter().any(|peer| peer.pk == node.pk)
+    }
+
+    fn is_iterate_time(&self, iterate_interval: Duration) -> bool {
+        self.last_time_send_ping.elapsed() >= iterate_interval
+    }
+
+    /// try to add node to list to send PingRequest
+    /// return true if node is added, false otherwise
+    pub fn try_add(&mut self, server: &Server, node: &PackedNode) -> bool {
+        // if node already exists in close list and not timed out, then don't send PingRequest
+        match server.close_nodes.read().find_node(&node.pk) {
+            Some(ref node_in_close_list) if !Ping::is_bad_node_timed_out(node_in_close_list, server) => return false,
+            _ => {},
+        };
+
+        // if node is not addable to close list, dont't send PingRequest
+        if !server.close_nodes.read().can_add(node) {
+            return false
+        }
+
+        // If node is friend and don't exist in friend's close list then send PingRequest
+        if Ping::is_friend(node, server) && !Ping::is_in_close_list(node, server) {
+            server.send_ping_req(node);
+            return false
+        }
+
+        let mut nodes_to_send_ping = self.nodes_to_send_ping.write();
+        // if node already exists in ping list, then don't add
+        if Ping::is_in_ping_list(node, nodes_to_send_ping.deref()) {
+            return false
+        }
+
+        // PingRequest is sent only for maximum 8 nodes in Bucket
+        nodes_to_send_ping.try_add(&server.pk, node)
+    }
+
+    /// send PingRequest to all nodes in list
+    pub fn iterate_ping_list(&mut self, server: &Server, iterate_interval: Duration) -> IoFuture<()> {
+        if !self.is_iterate_time(iterate_interval) {
+            return Box::new(future::ok(()))
+        }
+
+        let mut now = Instant::now();
+        let mut nodes_to_send_ping = Bucket::new(None);
+
+        mem::swap(&mut self.last_time_send_ping, &mut now);
+        mem::swap(self.nodes_to_send_ping.write().deref_mut(), &mut nodes_to_send_ping);
+
+        if nodes_to_send_ping.is_empty() {
+            return Box::new(future::ok(()))
+        }
+
+        let ping_sender = nodes_to_send_ping.nodes.iter().map(|node| {
+            server.send_ping_req(&(node.clone()).into())
+        });
+
+        let pings_stream = stream::futures_unordered(ping_sender).then(|_| Ok(()));
```

Comment:
Why `stream::futures_unordered` but not `future::join_all`?

Reply:
I think it is just my coding style. Is better to use join_all?

Reply:
`futures_unordered` returns a stream, but you actually don't need it, you need just a future.

Reply:
I used this iteration pattern for a long time. 
And there are many occurrences in DHT using this coding pattern.
Now, I got what you say.
Maybe, I may change all of this coding pattern in another PR.

Reply:
See https://docs.rs/futures/0.1.18/src/futures/stream/futures_unordered.rs.html#56 vs http://rust-lang-nursery.github.io/futures-rs/futures/future/fn.join_all.html .

Better use futures_unordered